### ISSUE TITLE
Support multinode training on GPU

### ISSUE THREAD

**Issue by py4** [google/jax#2731]
I don't have a node with 8 gpus. I have two nodes each with 4 gpus. So is it possible to train a model on multiple nodes?

**Comment by hawkinsp** [comment]
This *is* actually something that does work right now but it's still experimental. There's also no real public-facing API for it yet; you have to type in some obscure and fairly magical things to set it all up correctly.

We should polish it off and document it!

**Comment by hawkinsp** [comment]
Can you say a bit more about your model, though? Would gradient all-reductions across multiple nodes suffice?

**Comment by py4** [comment]
@hawkinsp Technically, I'm training a reformer model using Trax library.

**Comment by hawkinsp** [comment]
And I assume you're just looking for data parallelism, i.e., partitioning a minibatch across GPUs, not partitioning in any other way (e.g., model parallelism)?

**Comment by py4** [comment]
@hawkinsp yeah my concern is data parallelism

**Comment by powderluv** [comment]
@hawkinsp Can you please share your notes on this (don't need a stable api) ? We are trying some hybrid data/model/pipeline parallelism  so it is a little different from @py4 but would love to get started with data parallelism

**Comment by brettkoonce** [comment]
Data parallelism would of value to other projects that use XLA as well (eg https://www.tensorflow.org/swift).  Exposing this functionality in a standardized way would help drive progress in the broader ecosystem!

**Comment by yxd886** [comment]
> I don't have a node with 8 gpus. I have two nodes each with 4 gpus. So is it possible to train a model on multiple nodes?

Hello py4, I am meeting the same problem, have you found some solutions?

**Comment by yxd886** [comment]
> This _is_ actually something that does work right now but it's still experimental. There's also no real public-facing API for it yet; you have to type in some obscure and fairly magical things to set it all up correctly.
> 
> We should polish it off and document it!

Hello hawkinsp, Could you please provide more details about how to run data parallel with multi node GPUs?

**Comment by connection-on-fiber-bundles** [comment]
@hawkinsp We are also interested in running JAX code on multiple nodes. Anything (hacky or not) that you can share would be appreciated. Thanks!

**Comment by jramapuram** [comment]
I really enjoyed Jax during my DM internship and wanted to use it on my university SLURM cluster, but the lack of a clear (official) data parallel solution is a huge blocker to increasing Jax adoption outside of Google where you cant just grab a TPU pod. A single 8 (GPU) replica setup can barely train a Resnet50 imagenet classifier. Training SimCLR or any other large SOTA model is currently impossible without multi-node data parallelism.

**Comment by StellaAthena** [comment]
I would love this feature! I enjoy Jax, but I've been largely using DeepSpeed due to its ability to distribute across clusters.

**Comment by jrabary** [comment]
Any progress on this issue ? Using JAX to train a model on multi-node, multi-GPU is becoming a very important features for us.

**Comment by sudhakarsingh27** [comment]
This is a significant bottleneck for scaling on multi-node GPU clusters. Is there any update on this issue? 
Also, there was a recent `pjit` [tutorial](https://jax.readthedocs.io/en/latest/jax-101/08-pjit.html) that explains multi-node TPU scaling but doesn't mention about GPUs. Is that planned to be updated in the future?

**Comment by cloudhan** [comment]
@sudhakarsingh27 I constantly monitoring the jax releases, and there is something WIP that you might be interested in https://github.com/google/jax/pull/8364

**Comment by brettkoonce** [comment]
See also: #9582

**Comment by hawkinsp** [comment]
Yes indeed. We haven't advertised it that much yet, but (a) you need to initialize the cluster using that API, and (b) you need to follow the same rules of multi-host programming that also apply on TPU, documented here: https://jax.readthedocs.io/en/latest/multi_process.html

I suspect we can consider this issue closed when we've documented (a) in the document (b).

**Comment by sudhakarsingh27** [comment]
@hawkinsp @zhangqiaorjc 
Multinode (or multiprocess) doesn't seem to work with the following jax(lib) versions:
```
jax                           0.3.13                                                                                                                                                                       
jaxlib                        0.3.10+cuda11.cudnn82
```
Ran [the attached](https://github.com/google/jax/files/8744863/jax_multi_node_experiment.zip) minimal code on single node with 8 V100 GPUs as follows (2 processes with 4 GPUs each):

```
CUDA_VISIBLE_DEVICES="0,1,2,3" python jax_multi_node_experiment.py 0 &
CUDA_VISIBLE_DEVICES="4,5,6,7" python jax_multi_node_experiment.py 1
```
